### PR TITLE
chore(benchmarks): added benchmark code for agility cms.

### PR DESCRIPTION
**Pull request by joelvarty**
### Description

This adds a benchmarking setup for Agility CMS

I will send environment variables in another channel.

### CODE REVIEW

**Comment by pvdz** [comment]
@joelvarty Thanks for the benchmark!

I'll try do a run and check tomorrow (provided I can get my hands on the env vars), so just a pre-emptive question: can the benchmark be setup such a way that it pulls down the data once and reads from cache in other runs? For the benchmarks I'm less interested in the time it takes to download remote data so I prefer to not include that in the benchmark time. A quick glance through the readme implies it but I'm not sure, hence my question :)

Ideally it would be able to benchmark a custom number of pages (1k,10,100k), but that's probably tricky in general, and especially in combination with the caching.

**Comment by joelvarty** [comment]
Hi Peter - 

I've sent Alex a zip with 4 different env file combinations for the different site types.  
The first time the site runs it will pull the content into the GraphQL store and then on subsequent builds it will only pull the data that's been changed in Agility CMS.

Use `npm run update` in development mode to trigger a change in the content.

Hope that works - let me know if you need this to work differently.

File: benchmarks/source-agilitycms/package.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  "license": "MIT",
+  "scripts": {
+    "build": "gatsby clean && gatsby build",
+    "develop": "gatsby develop",
+    "format": "prettier --write src/**/*.{js,jsx}",
+    "start": "npm run develop",
+    "serve": "gatsby serve",
+    "test": "echo \"Write tests! -> https://gatsby.dev/unit-testing\"",
+	"setup": "NODE_ENV=development node bin/setup.js",
+	"update": "NODE_ENV=development node bin/update.js"
```
Comment (duffn) [inline benchmarks/source-agilitycms/package.json:15]: Please make this `data-update` to match the pattern used with other benchmarks, ex. https://github.com/gatsbyjs/gatsby/blob/947492971d33a508ace63f24c0be20870f3ae892/benchmarks/source-kontent/package.json#L13

File: benchmarks/source-agilitycms/gatsby-config.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+require("dotenv").config({
+  path: `.env.${process.env.NODE_ENV}`,
+})
+
+//configure your agility plugin with environment variables so that
+//your agility api credentials stay secure
+const agilityConfig = {
+  guid: process.env.AGILITY_GUID,
```
Comment (duffn) [inline benchmarks/source-agilitycms/gatsby-config.js:8]: Can you please prefix all environment variables with `BENCHMARK_`? This helps us easily identify benchmark related environment variables and conforms with the other benchmarks, ex. https://github.com/gatsbyjs/gatsby/blob/947492971d33a508ace63f24c0be20870f3ae892/benchmarks/source-kontent/gatsby-config.js#L17

**Comment by joelvarty** [comment]
Hi @duffn - I have made the changes you suggested about.  I have also emailed Alex with an updated file with all the various env files and values init.

File: benchmarks/source-agilitycms/src/agility/utils.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
+// const getDynamicPageItem = ({ contentID, agilityItem }) => {
```
Comment (duffn) [inline benchmarks/source-agilitycms/src/agility/utils.js:1]: Can we clean up these comments?

File: benchmarks/source-agilitycms/src/agility/components/AgilityPageTemplate.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+
+const AgilityPageTemplate = (props) => {
+	const pageTemplateName = props.page.templateName.replace(/[^0-9a-zA-Z]/g, '');
+	const PageTemplateComponentToRender = require(`../../pageTemplates/${pageTemplateName}.js`).default;
+
+	//get the page template name that we need to render
+
+	//const PageTemplateComponentToRender = props.pageTemplates[pageTemplateName];
```
Comment (duffn) [inline benchmarks/source-agilitycms/src/agility/components/AgilityPageTemplate.js:9]: Does this comment need removing?

File: benchmarks/source-agilitycms/package.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "agility-gatsby-starter",
+  "private": true,
+  "description": "A simplified bare-bones starter for Gatsby",
+  "version": "2.3.0",
+  "license": "MIT",
+  "scripts": {
+    "build": "gatsby clean && gatsby build",
```
[the remark this answers is not in the document]
Comment (joelvarty) [reply benchmarks/source-agilitycms/package.json]: Done!

File: benchmarks/source-agilitycms/gatsby-config.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        }],
+         //the page template that will be used to render Agility CMS pages
+        masterPageTemplate: "./src/AgilityPage.js"
+      },
+    },
+    `gatsby-plugin-sharp`,
+    `gatsby-transformer-sharp`,
+    {
+      resolve: `gatsby-source-filesystem`,
+      options: { name: `images`, path: `./src/img/` }
```
[the remark this answers is not in the document]
Comment (joelvarty) [reply benchmarks/source-agilitycms/gatsby-config.js]: Added the folder and switched that path to be `${__dirname}/src/img/`

File: benchmarks/source-agilitycms/src/agility/utils.js (hunk 3 above)
Comment (joelvarty) [reply benchmarks/source-agilitycms/src/agility/utils.js]: Cleaned!

File: benchmarks/source-agilitycms/src/agility/components/AgilityPageTemplate.js (hunk 4 above)
Comment (joelvarty) [reply benchmarks/source-agilitycms/src/agility/components/AgilityPageTemplate.js]: Indeed it did. Thanks!

File: benchmarks/source-agilitycms/gatsby-config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (joelvarty) [reply benchmarks/source-agilitycms/gatsby-config.js]: @duffn Should be good to go now!

[the remark this answers is not in the document]
Comment (duffn) [reply benchmarks/source-agilitycms/gatsby-config.js]: @joelvarty I'm still getting an error here. 

```
found 0 vulnerabilities
⠴ source and transform nodes
AgilityCMS => Content Sync returned 527 item(s).
AgilityCMS => Pulling Page Changes...
AgilityCMS => Page Sync returned 4 item(s).
AgilityCMS => Updating Sitemap channel website in en-us
AgilityCMS => Updating URL Redirections from lastAccessDate: null
AgilityCMS => Saving URL Redirections into cache
AgilityCMS => Completed Sync for en-us
AgilityCMS => Creating sitemap nodes...
AgilityCMS => Source Nodes Completed.

 ERROR


The path passed to gatsby-source-filesystem does not exist on your file system:
undefined
Please pick a path to an existing directory.
See docs here - https://www.gatsbyjs.org/packages/gatsby-source-filesystem/


not finished source and transform nodes - 73.008s
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! agility-gatsby-starter@2.3.0 build: `gatsby build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the agility-gatsby-starter@2.3.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

[the remark this answers is not in the document]
Comment (joelvarty) [reply benchmarks/source-agilitycms/gatsby-config.js]: Fixed!  I had a typo in the config file.  Thanks for catching that!

**Comment by joelvarty** [comment]
> I've testing building 512 # of pages locally and they work. We'll need @pvdz or @smthomas to approve and merge.

Awesome!  Can't wait to get this live!

File: benchmarks/source-agilitycms/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  </a>
+</p>
+<h1 align="center">
+  Gatsby + Agility CMS : WILL IT BUILD
+</h1>
+
+## To Update Content from the CMS
+
+```shell
+num run update
```
Comment (pvdz) [inline benchmarks/source-agilitycms/README.md:14]: Is this a typo? I'm not sure what `nvm` is supposed to be here, but probably not node version manager?

In the package.json I do see a `data-update` so perhaps this was meant to be `npm data-update`?

Also, when I try to run that it says 
> An error occurred while updating the posts list:  TypeError: You must provide either guid or a websiteName.

Not sure how to proceed from there or whether it's meant for consumption outside of Agility CMS-staff at all.

File: benchmarks/source-agilitycms/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+```
+
+## Pull Content From Agility CMS
+
+For the first time setup only (this has been done by Agility CMS staff)
+
+```shell
+npm run setup
+```
+
```
Comment (pvdz) [inline benchmarks/source-agilitycms/README.md:24]: While a little overkill, I think it'd be good to be explicit in how to build, even if it's "just the default". This prevents an engineer potentially overthinking it, looking up to make sure that it needs no other args.

````suggestion
## Build

```shell
gatsby build
```
````

**Comment by pvdz** [comment]
I ran the first benchmark ("1") on my machine. That's 500 pages. 

The very first run:

Roughly speaking, the sourcing step completed in 430s (7min). Webpack JS bundle time is 14s. Run queries 30s. 

Apparently the site rocks 2900 images. In a cold build, those take a long time locally, around 440s (7min).

Total time was 870s, but 800 of that was sourcing and images (the less relevant part of a benchmark as far as I'm concerned as Gatsby has no or almost no control over how fast they completes).

The second run (warm):

Finished in roughly 10s. Nice

Third build (cold, so `gatsby clean; gatsby build`):

I think this skipped an update step which it had to do for the very first time after cloning so it completed a little faster this time:

Sourcing 262s, webpack 15s, run queries 41s, images 508s.

Anyways, seems to work here :)

File: benchmarks/source-agilitycms/README.md (hunk 7 above)
Comment (joelvarty) [reply benchmarks/source-agilitycms/README.md:14]: It was totally a typo.  If you run `npm run data-update` it will update content in the CMS.

**Comment by joelvarty** [comment]
Thanks for the review @pvdz! Made the tweak to the readme.md file.  Let me know if you think we need to make any other changes here.

**Comment by pvdz** [comment]
I think it's good to go. I ran the formatter to fix those linting messages. There aren't any other tests around benchmarks afaik.

**Pull request merged**